[이주명] Mission 3: 회원가입 기능 구현

jumyeong33 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 회원가입  views.py 작성 
- lib 디렉토리 생성 
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 회원가입 : 이름, 비밀번호, 핸드폰번호, 이메일, 설명
- email validation and password validation 함수 생성
- 비밀번호 형식 실패시 에러 반환
- 이메일 중복시 에러 반환 
- 이메일 형식 실패시 에러 반환
- 이메일 or 비밀번호 미기재시 에러 반환 
- 성공시 201 CREATED 반환

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
 class User() 를 만들어서 각 view 마다 재사용될 인자들을 받으려면 어떻게 테스트 해야되나요? 

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

jumyeong33 committed:
change '/' in urls.py

jumyeong33 committed:
rebase mode1

jumyeong33 committed:
order of data

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,31 @@
-# Create your views here.
+from django.http import JsonResponse
+from django.views import View
+from users.models import User
```
장고 import 문과 구별되게 한칸 정도 공백을 두시는걸 추천드립니다!~

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,31 @@
+from users.models import User
+from users.lib.validation import email_validation, password_validation
+
+class UserView(View) :
```
웹 개발시 클래스 네이밍은 보다 기능을 명확히 알 수 있는 클래스 명으로 정해주셔야 합니다.  SignUpView 정도로 변경해주세요!~

kyeongraekim commented on students/jumyoeng/users/urls.py:
```diff
@@ -0,0 +1,7 @@
+from users.views import UserView
+
+urlpatterns = [
+    path('/create', UserView.as_view())
```
Create 는 너무 포괄적인 url 주소로 보여집니다. /signup 정도로 유저정보의 회원가입을 의미하는 내용으로 변경해주세요!~

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,31 @@
+            name         = data['name'],
+            email        = data['email'],
+            password     = data['password'],
+            phone_number = data['phone'],
```
phone_number 는 모델상 null=True 로 반영되어있네요? null 값이 허용되는 데이터는 딕셔너리 매핑 메소드중 .get()을 사용하여 null을 받을 수 있게 변경해보세요!~

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,31 @@
+            )
+            return JsonResponse({'MESSAGE': 'CREATED'}, status=201)
+
+        except KeyError:
```
위 코드 진행시 발생할 수 있는 키에러 말고 다른 여러가지가 있을 수 있을거라 생각됩니다. 개발자의 시선에서 어떠한 에러들을 핸들링해야할지 한 번 고민해주세요!~

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,31 @@
+            if User.objects.filter(email=data['email']) :
+                return JsonResponse({'MESSAGE': 'EMAIL_DUPLICATE'}, status=400)
+
+            if not email_validation(data['email']) :
+                return JsonResponse({'MESSAGE': 'EMAIL_INVALID'}, status=400)
+
+            if not password_validation(data['password']) :
+                return JsonResponse({'MESSAGE': 'PASSWORD_INVALID'}, status=400)
```
해당 validation 파일들이 이번 브랜치 선에서 보이지가 않네요~ 직접 모듈로서 작성하신 파일이라면 깃허브에 푸시를 해주시고, 따로 만들어진 라이브러리를 사용하신거라면 보다 장고를 퓨어하게 사용해볼 수 있게 직접 작성해보세요!~

jumyeong33 replied:
lib 디렉토리를 따로 만들었는데 gitignore에 lib이 명시되어 있네요! 수정했습니다.  :)

kyeongraekim reviewed (changes requested):
아래 내용 참고해서 수정해주세요!~

jumyeong33 committed:
Fixed up

-  change urls.py
- change view.py class name
- fix phone_number could get null, as default
- add except Exception fo other errors

jumyeong33 committed:
validation moduel for email and password

jumyeong33 committed:
Wrong import

kyeongraekim commented on students/jumyoeng/users/my_lib/validation.py:
```diff
@@ -0,0 +1,26 @@
+import re 
+
+def email_validation(email) :
+    em_form =r'^[a-z0-9]+[\._]?[a-z0-9]+[@]\w+[.]\w{2,3}$'
+    vaild = re.search(em_form , email)
+    if vaild : 
+        return True
+    else :
+        return False
+
+def password_validation(password) :
+    if len(password) < 8 :
+        print("최소 8자 이상 입력해주세요.")
+        return False
+    elif re.search('[\d]+', password) is None:
+        print("최소 하나의 숫자가 포함되어야 합니다.")
+        return False
+    elif re.search('[\D]+', password) is None:
+        print("최소 하나의 문자가 포함되어야 합니다.")
+        return False
+    elif re.search('[,./~!@#$%^&*()><]+', password) is None:
+        print("최소 하나의 특수문자가 포함되어야 합니다.")
+        return False
+    else :
+        return True
+ 
```
해당 내용은 4단계부터 전반적으로 리뷰 달아드릴 내용임으로 Mission3 완료 후 수정 내용 달아드리겠습니다!~

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,36 @@
+            email        = data['email'],
+            password     = data['password'],
+            phone_number = None if data.get('phone') is None else data['phone'],
+            description  = data['descr'],
```
프론트와 통신을 주고 받을때에도 영어 원어를 써 주시는것을 권장드립니다!

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,36 @@
+            name         = data['name'],
+            email        = data['email'],
+            password     = data['password'],
+            phone_number = None if data.get('phone') is None else data['phone'],
```
3항연산자를 굳이 쓸 필요 없이 data.get('phone') 에 해당되는 매핑이 불가할 시 none 이 디폴트로 나오게 됩니다.

kyeongraekim commented on students/jumyoeng/users/views.py:
```diff
@@ -1,3 +1,36 @@
+        except KeyError:
+            return  JsonResponse({'MESSAGE': 'KEY_ERROR'}, status=400)
+
+        except Exception as e :
```
깃허브에 올라오는 개발 코드는 개인 디버깅에 쓰일만한 print 문들은 모두 지워주시고 푸시해주셔야합니다!~

kyeongraekim reviewed (changes requested):
아래내용 확인하여 수정해주세요!~

jumyeong33 committed:
fixed up [2]

- delete print()
- get json data first

jumyeong33 committed:
last modified

kyeongraekim reviewed (approved):
LGTM!~
